packages/logging/service.go

```diff
@@ -31,6 +36,52 @@ type loggingService struct {
 	logProvider logProvider
 }
 
+// QueryForDogu writes dogu log messages into the stream of the given server.
+func (s *loggingService) QueryForDogu(request *pb.DoguLogMessageQueryRequest, server pb.DoguLogMessages_QueryForDoguServer) error {
+	doguName := request.DoguName
+	if doguName == "" {
+		return status.Error(codes.InvalidArgument, responseMessageMissingDoguname)
+	}
+
+	var filter = ""
+	if request.Filter != nil {
+		filter = request.GetFilter()
+	}
+
+	var startDate *time.Time = nil
+	if request.GetStartDate() != nil {
+		date := request.GetStartDate().AsTime()
+		startDate = &date
+	}
+
+	var endDate *time.Time = nil
+	if request.GetEndDate() != nil {
+		date := request.GetEndDate().AsTime()
+		endDate = &date
+	}
+
+	logrus.Debugf("retrieving log messages from %s to %s for dogu '%s' with filter %s", startDate, endDate, doguName, filter)
+
+	logLines, err := s.logProvider.queryLogs(doguName, startDate, endDate, filter)
+	if err != nil {
+		logrus.Errorf("error reading logs: %v", err)
+		return createInternalErr(err, codes.InvalidArgument)
+	}
+
+	for _, line := range logLines {
+		err := server.Send(&pb.DoguLogMessage{
+			Timestamp: timestamppb.New(line.timestamp),
+			Message:   line.value,
+		})
+		if err != nil {
+			logrus.Errorf("error writing log-lines to stream: %v", err)
+			return createInternalErr(err, codes.InvalidArgument)
+		}
+	}
```

Comment:
Maybe it is better to run queryLogs in a goroutine and return the logs via a channel? By using the channel you can send the logs for every request you made to loki. At the moment, you are batching all logs in k8s-ces-control and then send them via a stream, which doesn't make use of the advantages of the stream.

Reply:
Since we have to load the logs from Loki sequentially I see now advantage in using go routines either.
We can not make concurrent requests at Loki, because we need the date of the last element from an response for the new endDate for the next request.